File: src/utils/finance/adwin.ts
/**
 * Mirrors `research/models/adwin.py`.
 *
 * ADWIN (ADaptive WINdowing) drift detector.
 *
 * Bifet & Gavaldà 2007 — "Learning from Time-Changing Data with Adaptive
 * Windowing". Maintains a variable-length window of recent observations,
 * compressed into exponentially-sized buckets, and shrinks the window
 * whenever a Hoeffding-bound test detects a change in the mean between
 * any prefix and the corresponding suffix.
 *
 * Use as a "train-only-when-required" trigger for adaptive estimators.
 *
 * Empirical reference: Yelleti 2025 (arXiv 2504.10229) — ADWIN won 4/5
 * datasets in the ROSFD streaming-fraud benchmark vs DDM/EDDM.
 */

interface AdwinOptions {
  /** Per-test false-positive rate. Smaller => fewer false alarms, slower detect. */
  delta?: number;
  /** Max buckets per row before compaction (M in the paper). */
  maxBuckets?: number;
  /** Min total window size before drift checks fire. */
  minWindow?: number;
}

interface Bucket {
  total: number; // sum of values in this bucket
  variance: number; // sum of (xi - bucketMean)^2 within this bucket
  size: number; // number of samples
}

interface Row {
  buckets: Bucket[]; // buckets at this exponential level
}

const DEFAULT_DELTA = 0.002;
const DEFAULT_MAX_BUCKETS = 5;
const DEFAULT_MIN_WINDOW = 10;

export class Adwin {
  private rows: Row[] = [];
  private total = 0;
  private variance = 0; // sum of (xi - mean)^2 over the window (Welford-style)
  private width = 0;
  private readonly delta: number;
  private readonly maxBuckets: number;
  private readonly minWindow: number;
  private lastDrift = false;

  constructor(opts: AdwinOptions = {}) {
    this.delta = opts.delta ?? DEFAULT_DELTA;
    this.maxBuckets = opts.maxBuckets ?? DEFAULT_MAX_BUCKETS;
    this.minWindow = opts.minWindow ?? DEFAULT_MIN_WINDOW;
  }

  size(): number {
    return this.width;
  }

  mean(): number {
    if (this.width === 0) return Number.NaN;
    return this.total / this.width;
  }

  driftDetected(): boolean {
    return this.lastDrift;
  }

  bucketCount(): number {
    let c = 0;
    for (const r of this.rows) c += r.buckets.length;
    return c;
  }

  /** Push a new observation; returns true if drift was detected on this step. */
  add(value: number): boolean {
    this.insert(value);
    this.compress();
    this.lastDrift = this.checkDrift();
    return this.lastDrift;
  }

  private insert(value: number): void {
    if (this.rows.length === 0) this.rows.push({ buckets: [] });
    const row0 = this.rows[0];
    row0.buckets.push({ total: value, variance: 0, size: 1 });
    if (this.width > 0) {
      const oldMean = this.total / this.width;
      const incr = ((value - oldMean) * (value - oldMean) * this.width) / (this.width + 1);
      this.variance += incr;
    }
    this.total += value;
    this.width += 1;
  }

  private compress(): void {
    let level = 0;
    while (level < this.rows.length) {
      const row = this.rows[level];
      if (row.buckets.length <= this.maxBuckets) break;
      // Merge the two oldest buckets (front of the array) and promote.
      const b1 = row.buckets.shift()!;
      const b2 = row.buckets.shift()!;
      const newSize = b1.size + b2.size;
      const newTotal = b1.total + b2.total;
      const newMean = newTotal / newSize;
      const m1 = b1.total / b1.size;
      const m2 = b2.total / b2.size;
      const newVar =
        b1.variance + b2.variance + (b1.size * b2.size * (m1 - m2) * (m1 - m2)) / newSize;
      const merged: Bucket = { total: newTotal, variance: newVar, size: newSize };
      if (level + 1 >= this.rows.length) this.rows.push({ buckets: [] });
      this.rows[level + 1].buckets.push(merged);
      level += 1;
    }
  }

  private checkDrift(): boolean {
    if (this.width < this.minWindow) return false;
    let detected = false;
    let dropAttempts = 0;
    const MAX_DROPS = 5;
    while (dropAttempts < MAX_DROPS) {
      let droppedThisPass = false;
      // Walk buckets from oldest to newest; compute prefix sum and test.
      let n0 = 0;
      let s0 = 0;
      let v0 = 0;
      const flatBuckets: { bucket: Bucket; row: number; idx: number }[] = [];
      for (let r = this.rows.length - 1; r >= 0; r--) {
        for (let i = 0; i < this.rows[r].buckets.length; i++) {
          flatBuckets.push({ bucket: this.rows[r].buckets[i], row: r, idx: i });
        }
      }
      for (let k = 0; k < flatBuckets.length - 1; k++) {
        const b = flatBuckets[k].bucket;
        n0 += b.size;
        s0 += b.total;
        v0 += b.variance;
        const n1 = this.width - n0;
        if (n0 < this.minWindow / 2 || n1 < this.minWindow / 2) continue;
        const mean0 = s0 / n0;
        const mean1 = (this.total - s0) / n1;
        // Hoeffding bound (Bifet & Gavaldà 2007 eq. for ε_cut)
        const m = 1 / (1 / n0 + 1 / n1);
        const dPrime = this.delta / Math.log(this.width);
        const varEst = this.variance / this.width;
        const epsilon =
          Math.sqrt((2 / m) * varEst * Math.log(2 / dPrime)) +
          (2 / 3) * Math.log(2 / dPrime) / m;
        if (Math.abs(mean0 - mean1) > epsilon) {
          // Drop the oldest bucket and recompute.
          const oldest = flatBuckets[0];
          const dropped = this.rows[oldest.row].buckets.shift()!;
          this.total -= dropped.total;
          this.width -= dropped.size;
          // Recompute variance from scratch (cheap; window is bounded).
          this.recomputeVariance();
          detected = true;
          droppedThisPass = true;
          break;
        }
      }
      if (!droppedThisPass) break;
      dropAttempts++;
    }
    return detected;
  }

  private recomputeVariance(): void {
    if (this.width === 0) {
      this.variance = 0;
      return;
    }
    const mean = this.total / this.width;
    let v = 0;
    for (const row of this.rows) {
      for (const b of row.buckets) {
        const bMean = b.total / b.size;
        v += b.variance + b.size * (bMean - mean) * (bMean - mean);
      }
    }
    this.variance = v;
  }
}

File: research/models/adwin.py
"""ADWIN (ADaptive WINdowing) drift detector — mirror of TS adwin.ts.

Bifet & Gavaldà 2007.  Maintains a variable-length window of recent
observations, compressed into exponentially-sized buckets, and shrinks
the window whenever a Hoeffding-bound test detects a change in mean
between any prefix and corresponding suffix.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List


@dataclass
class _Bucket:
    total: float
    variance: float
    size: int


@dataclass
class _Row:
    buckets: List[_Bucket] = field(default_factory=list)


class Adwin:
    def __init__(
        self,
        delta: float = 0.002,
        max_buckets: int = 5,
        min_window: int = 10,
    ) -> None:
        self.delta = delta
        self.max_buckets = max_buckets
        self.min_window = min_window
        self._rows: List[_Row] = []
        self._total = 0.0
        self._variance = 0.0
        self._width = 0
        self._last_drift = False

    def size(self) -> int:
        return self._width

    def mean(self) -> float:
        if self._width == 0:
            return float("nan")
        return self._total / self._width

    def drift_detected(self) -> bool:
        return self._last_drift

    def bucket_count(self) -> int:
        return sum(len(r.buckets) for r in self._rows)

    def add(self, value: float) -> bool:
        self._insert(value)
        self._compress()
        self._last_drift = self._check_drift()
        return self._last_drift

    def _insert(self, value: float) -> None:
        if not self._rows:
            self._rows.append(_Row())
        self._rows[0].buckets.append(_Bucket(total=value, variance=0.0, size=1))
        if self._width > 0:
            old_mean = self._total / self._width
            incr = (value - old_mean) ** 2 * self._width / (self._width + 1)
            self._variance += incr
        self._total += value
        self._width += 1

    def _compress(self) -> None:
        level = 0
        while level < len(self._rows):
            row = self._rows[level]
            if len(row.buckets) <= self.max_buckets:
                break
            b1 = row.buckets.pop(0)
            b2 = row.buckets.pop(0)
            new_size = b1.size + b2.size
            new_total = b1.total + b2.total
            m1 = b1.total / b1.size
            m2 = b2.total / b2.size
            new_var = (
                b1.variance + b2.variance + (b1.size * b2.size * (m1 - m2) ** 2) / new_size
            )
            merged = _Bucket(total=new_total, variance=new_var, size=new_size)
            if level + 1 >= len(self._rows):
                self._rows.append(_Row())
            self._rows[level + 1].buckets.append(merged)
            level += 1

    def _check_drift(self) -> bool:
        if self._width < self.min_window:
            return False
        detected = False
        drop_attempts = 0
        max_drops = 5
        while drop_attempts < max_drops:
            dropped_this_pass = False
            n0 = 0
            s0 = 0.0
            flat = []
            for r in range(len(self._rows) - 1, -1, -1):
                for i, b in enumerate(self._rows[r].buckets):
                    flat.append((b, r, i))
            for k in range(len(flat) - 1):
                b = flat[k][0]
                n0 += b.size
                s0 += b.total
                n1 = self._width - n0
                if n0 < self.min_window / 2 or n1 < self.min_window / 2:
                    continue
                mean0 = s0 / n0
                mean1 = (self._total - s0) / n1
                m = 1 / (1 / n0 + 1 / n1)
                d_prime = self.delta / math.log(self._width)
                var_est = self._variance / self._width
                epsilon = math.sqrt((2 / m) * var_est * math.log(2 / d_prime)) + (
                    2 / 3
                ) * math.log(2 / d_prime) / m
                if abs(mean0 - mean1) > epsilon:
                    oldest_row = flat[0][1]
                    dropped = self._rows[oldest_row].buckets.pop(0)
                    self._total -= dropped.total
                    self._width -= dropped.size
                    self._recompute_variance()
                    detected = True
                    dropped_this_pass = True
                    break
            if not dropped_this_pass:
                break
            drop_attempts += 1
        return detected

    def _recompute_variance(self) -> None:
        if self._width == 0:
            self._variance = 0.0
            return
        mean = self._total / self._width
        v = 0.0
        for row in self._rows:
            for b in row.buckets:
                b_mean = b.total / b.size
                v += b.variance + b.size * (b_mean - mean) ** 2
        self._variance = v
